Add Bamba Model

fabianlim opened the pull request:
This is the companion PR to an [huggingface PR](https://github.com/huggingface/transformers/pull/34982) for adding `Bamba`, which is a hybrid [mamba2](https://github.com/state-spaces/mamba) architecture with SwiGLU. The checkpoints are jointly trained by IBM, Princeton, and UIUC.

In this PR we have:
- [x] Created the `bamba` model inference architecture, which we would like acknowledge the `jamba` team for referencing their implementation, whereby we modified to support full attention layers with RoPE and mamba v2.
- [x] Ensured that we have TP support.
- [x] Ensured we  support chunked prefill. ~Currently we have a partial solution, which works only when the cont batch boundaries line up with the chunked boundaries.~ This is now completely fixed.
- [x] Ensured that we conform to the recent [PR](https://github.com/vllm-project/vllm/pull/10992) for adding pipeline support for SSM models. 
- [x] Adapted the mamba v2 scan kernels into `vllm/model_executor/layers/mamba/ops`. Only the `fwd` kernels are extracted. Some modifications and fixes are made.
- [x] created `tests/models/decoder_only/language/test_bamba.py` with an initial `ibm-fms/Bamba-9.8b-1.8T-hf`. This is practically identical to `test_mamba.py`, only chunked prefill tests are disabled as it is currently not supported.

Currently only `FlashAttention` backend is supported, as we check fields like `context_lens_tensor`.  Have not yet investigated other backends.

We would like to also acknowledge the [draft codestral mamba PR](https://github.com/vllm-project/vllm/pull/9292) from @tlrmchlsmth, which we also referenced the mixer. 
- we made a few simplications for bamba (simplified mixer from mamba v2)
- Cuda graph capturing seems to be working, but we understand that cudagraphs are disabled for long sequence lengths. For SSM models the strength is in this regime, so can we handle it better?

Hope to discuss the following with the maintainers
1. ~do we have to remove all the `bwd` kernels?~ yes we should
2. for the full attention layers, we increase the `sin_cos` cache to cover the sequence length, if it is longer than `max_sequence_len`. ~This differs for other current models (e.g., llama). How can we better support long sequence lengths?~ we should keep this consistent with other models, so we propose to allow the `sin_cos` cache extension only when `VLLM_ALLOW_LONG_MAX_MODEL_LEN` is specified.
3. ~have some ideas to support chunked pre-fill, but will appreciate some discussion with the maintainers on how to proceed.~ working on changing the kernels to support chunked prefill.
4. ~since the mixer2 is simplified from mamba, should we rename it?~ we can keep it as is, but we should document the differences from `mamba_ssm`



cc: @ani300, @raghukiran1224, @cyang49, @njhill

fabianlim committed:
initial pr without tp fix

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
fix casting in rms norm gated

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
TP fix

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

tlrmchlsmth commented:
Hi @fabianlim, thanks for the PR! It's really great to see progress being made on state-space models, especially for me as I unfortunately haven't been able to prioritize support for Mamba2

I'm happy to shepherd this PR and discuss any questions you have, especially to support chunked prefill. If you haven't already, can you join the developer slack for quicker discussion? (https://communityinviter.com/apps/vllm-dev/join-vllm-developers-slack)

fabianlim committed:
fix mamba scan invalid address

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
some fixes and remove unused kernels

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
fmt + lint

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
more comments

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
initial fix for chunked prefill (incomplete)

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
improve comments

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
do not attach seq_idx to attn_metadata

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
activate initial states for chunked prefill

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim commented:
@tlrmchlsmth I cleaned up the PR quite abit, perhaps it might be a good time to get some early eyes.

tlrmchlsmth commented on tests/models/decoder_only/language/test_bamba.py:
```diff
@@ -0,0 +1,329 @@
+
+from ...utils import check_outputs_equal
+
+# will be ch
+MODELS = ["ibm-fms/Bamba-9.8b-1.8T-hf"]
```
The comment trails off, but will there be a small test model available?

fabianlim replied:
@raghukiran1224 any plans for a small test model? I think since we do outputs comparison it is not that good to just have a randomly initialised small model

raghukiran1224 replied:
@fabianlim @tlrmchlsmth would it be ok to test with a random model or would you rather have a tiny model (say 200M or so) to test with?

tlrmchlsmth replied:
A tiny model with nonrandom weights would be much better!

tlrmchlsmth replied:
btw is there any update on this?

tlrmchlsmth commented on tests/models/decoder_only/language/test_bamba.py:
```diff
@@ -0,0 +1,329 @@
+MODELS = ["ibm-fms/Bamba-9.8b-1.8T-hf"]
+
+
+# Use lower-level interfaces to create this greedy generator, as mamba will
+# choke on the model_kwarg 'attention_mask' if hf_model.generate_greedy is used.
```
Is this actually true of Bamba? I think this might not be the case for hybrid models, only attention-free ones.

We could consider renaming `tests/models/decoder_only/language/test_jamba.py` to `tests/models/decoder_only/language/test_hybrid.py` and then extending it to support bamba as well. Alternatively we could adapt the `test_jamba.py` tests to test bamba instead of adapting `test_mamba.py` tests.

fabianlim replied:
@tlrmchlsmth ah ok.. thats a good point.. let me look at `test_jamba.py`

fabianlim replied:
I did a quick test on the `test_jamba.py`, everything passed except these 3 below
- as expected prefill chunking fails for now until the WIP is resolved.
- `test_parallel_sampling` we seem to require `float` to pass the similar test in `test_bamba.py`
- for the last one Im abit confused. the stacktrace was `Cannot re-initialize CUDA in forked subprocess. To use CUDA with multiprocessing, you must use the 'spawn' start method`

<details>
Traceback (most recent call last):
  File "/data/flim/vllm/tests/utils.py", line 658, in wrapper
    f(*args, **kwargs)
  File "/data/flim/vllm/tests/models/decoder_only/language/test_jamba.py", line 328, in test_jamba_distributed_produces_identical_generation
    with vllm_runner(model, dtype=dtype, tensor_parallel_size=2) as vllm_model:
  File "/data/flim/vllm/tests/conftest.py", line 672, in __init__
    self.model = LLM(
  File "/data/flim/vllm/vllm/utils.py", line 1063, in inner
    return fn(*args, **kwargs)
  File "/data/flim/vllm/vllm/entrypoints/llm.py", line 232, in __init__
    self.llm_engine = self.engine_class.from_engine_args(
  File "/data/flim/vllm/vllm/engine/llm_engine.py", line 575, in from_engine_args
    engine = cls(
  File "/data/flim/vllm/vllm/engine/llm_engine.py", line 336, in __init__
    self.model_executor = executor_class(vllm_config=vllm_config, )
  File "/data/flim/vllm/vllm/executor/distributed_gpu_executor.py", line 26, in __init__
    super().__init__(*args, **kwargs)
  File "/data/flim/vllm/vllm/executor/executor_base.py", line 36, in __init__
    self._init_executor()
  File "/data/flim/vllm/vllm/executor/multiproc_gpu_executor.py", line 113, in _init_executor
    self._run_workers("init_device")
  File "/data/flim/vllm/vllm/executor/multiproc_gpu_executor.py", line 195, in _run_workers
    driver_worker_output = driver_worker_method(*args, **kwargs)
  File "/data/flim/vllm/vllm/worker/worker.py", line 138, in init_device
    torch.cuda.set_device(self.device)
  File "/workspace/mamba-vllm/lib/python3.10/site-packages/torch/cuda/__init__.py", line 478, in set_device
    torch._C._cuda_setDevice(device)
  File "/workspace/mamba-vllm/lib/python3.10/site-packages/torch/cuda/__init__.py", line 305, in _lazy_init
    raise RuntimeError(
RuntimeError: Cannot re-initialize CUDA in forked subprocess. To use CUDA with multiprocessing, you must use the 'spawn' start method

</details>

```
FAILED tests/models/decoder_only/language/test_jamba.py::test_mamba_prefill_chunking[10-bfloat16-ibm-fms/Bamba-9.8b-1.8T-hf] - AssertionError: Test0:
FAILED tests/models/decoder_only/language/test_jamba.py::test_parallel_sampling[15-bfloat16-ibm-fms/Bamba-9.8b-1.8T-hf] - AssertionError: Test0:
FAILED tests/models/decoder_only/language/test_jamba.py::test_jamba_distributed_produces_identical_generation[64-float-ibm-fms/Bamba-9.8b-1.8T-hf] - AssertionError: function <function test_jamba_distributed_produces_identical_generation at 0x7fbefda9a5f0> failed when called with args () and kwargs {'vllm_runner': <class 'tests.conftest.VllmRunner'>, 'model': 'i...
```

fabianlim replied:
Ah ok for multi-gpu tests you need to do `pytest -v -s` to avoid the fork error. The test passes!

<details>

Fork a new process to run a test 0
INFO 12-13 02:08:38 __init__.py:46] No plugins found.
INFO 12-13 02:08:55 config.py:403] This model supports multiple tasks: {'generate', 'embedding'}. Defaulting to 'generate'.
INFO 12-13 02:08:55 config.py:1038] Defaulting to use mp for distributed inference
INFO 12-13 02:08:55 llm_engine.py:249] Initializing an LLM engine (v0.1.dev3512+g388ee3d) with config: model='ibm-fms/Bamba-9.8b-1.8T-hf', speculative_config=None, tokenizer='ibm-fms/Bamba-9.8b-1.8T-hf', skip_tokenizer_init=False, tokenizer_mode=auto, revision=None, override_neuron_config=None, tokenizer_revision=None, trust_remote_code=True, dtype=torch.float32, max_seq_len=1024, download_dir=None, load_format=LoadFormat.AUTO, tensor_parallel_size=2, pipeline_parallel_size=1, disable_custom_all_reduce=False, quantization=None, enforce_eager=False, kv_cache_dtype=auto, quantization_param_path=None, device_config=cuda, decoding_config=DecodingConfig(guided_decoding_backend='xgrammar'), observability_config=ObservabilityConfig(otlp_traces_endpoint=None, collect_model_forward_time=False, collect_model_execute_time=False), seed=0, served_model_name=ibm-fms/Bamba-9.8b-1.8T-hf, num_scheduler_steps=1, chunked_prefill_enabled=False multi_step_stream_outputs=True, enable_prefix_caching=False, use_async_output_proc=True, use_cached_outputs=False, mm_processor_kwargs=None, pooler_config=None,compilation_config=CompilationConfig(level=0, backend='', custom_ops=[], splitting_ops=['vllm.unified_attention', 'vllm.unified_attention_with_output'], use_inductor=True, inductor_specialize_for_cudagraph_no_more_than=None, inductor_compile_sizes=[], inductor_compile_config={}, inductor_passes={}, use_cudagraph=False, cudagraph_num_of_warmups=0, cudagraph_capture_sizes=None, cudagraph_copy_inputs=False, pass_config=PassConfig(dump_graph_stages=[], dump_graph_dir=PosixPath('.'), enable_fusion=True, enable_reshape=True), compile_sizes=[], capture_sizes=[256, 248, 240, 232, 224, 216, 208, 200, 192, 184, 176, 168, 160, 152, 144, 136, 128, 120, 112, 104, 96, 88, 80, 72, 64, 56, 48, 40, 32, 24, 16, 8, 4, 2, 1], enabled_custom_ops=Counter(), disabled_custom_ops=Counter(), static_forward_context={})
WARNING 12-13 02:09:01 multiproc_gpu_executor.py:56] Reducing Torch parallelism from 40 threads to 1 to avoid unnecessary CPU contention. Set OMP_NUM_THREADS in the external environment to tune this value as needed.
INFO 12-13 02:09:01 custom_cache_manager.py:17] Setting Triton cache manager to: vllm.triton_utils.custom_cache_manager:CustomCacheManager
INFO 12-13 02:09:01 selector.py:222] Cannot use FlashAttention-2 backend for dtype other than torch.float16 or torch.bfloat16.
INFO 12-13 02:09:01 selector.py:129] Using XFormers backend.
(VllmWorkerProcess pid=1587462) INFO 12-13 02:09:01 selector.py:222] Cannot use FlashAttention-2 backend for dtype other than torch.float16 or torch.bfloat16.
(VllmWorkerProcess pid=1587462) INFO 12-13 02:09:01 selector.py:129] Using XFormers backend.
(VllmWorkerProcess pid=1587462) INFO 12-13 02:09:01 multiproc_worker_utils.py:215] Worker ready; awaiting tasks
INFO 12-13 02:09:02 utils.py:996] Found nccl from library libnccl.so.2
(VllmWorkerProcess pid=1587462) INFO 12-13 02:09:02 utils.py:996] Found nccl from library libnccl.so.2
INFO 12-13 02:09:02 pynccl.py:69] vLLM is using nccl==2.21.5
(VllmWorkerProcess pid=1587462) INFO 12-13 02:09:02 pynccl.py:69] vLLM is using nccl==2.21.5
INFO 12-13 02:09:02 custom_all_reduce_utils.py:242] reading GPU P2P access cache from /workspace/.cache/vllm/gpu_p2p_access_cache_for_0,1.json
(VllmWorkerProcess pid=1587462) INFO 12-13 02:09:02 custom_all_reduce_utils.py:242] reading GPU P2P access cache from /workspace/.cache/vllm/gpu_p2p_access_cache_for_0,1.json
INFO 12-13 02:09:02 shm_broadcast.py:236] vLLM message queue communication handle: Handle(connect_ip='127.0.0.1', local_reader_ranks=[1], buffer=<vllm.distributed.device_communicators.shm_broadcast.ShmRingBuffer object at 0x7f940c04f730>, local_subscribe_port=33527, remote_subscribe_port=None)
INFO 12-13 02:09:02 model_runner.py:1089] Starting to load model ibm-fms/Bamba-9.8b-1.8T-hf...
(VllmWorkerProcess pid=1587462) INFO 12-13 02:09:02 model_runner.py:1089] Starting to load model ibm-fms/Bamba-9.8b-1.8T-hf...
INFO 12-13 02:09:06 weight_utils.py:243] Using model weights format ['*.safetensors']
(VllmWorkerProcess pid=1587462) INFO 12-13 02:09:06 weight_utils.py:243] Using model weights format ['*.safetensors']
Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]
Loading safetensors checkpoint shards:  25% Completed | 1/4 [00:01<00:04,  1.40s/it]
Loading safetensors checkpoint shards:  50% Completed | 2/4 [00:02<00:02,  1.49s/it]
Loading safetensors checkpoint shards:  75% Completed | 3/4 [00:04<00:01,  1.53s/it]
Loading safetensors checkpoint shards: 100% Completed | 4/4 [00:06<00:00,  1.57s/it]
Loading safetensors checkpoint shards: 100% Completed | 4/4 [00:06<00:00,  1.54s/it]

INFO 12-13 02:09:14 model_runner.py:1094] Loading model weights took 18.4902 GB
(VllmWorkerProcess pid=1587462) INFO 12-13 02:09:16 model_runner.py:1094] Loading model weights took 18.4902 GB
(VllmWorkerProcess pid=1587462) INFO 12-13 02:10:28 worker.py:237] Memory profiling results: duration=71.88 seconds, total_gpu_memory=79.15GiB, initial_memory_usage=19.68GiB, peak_torch_memory=34.32GiB, memory_usage_post_profile=35.32GiB, non_torch_memory=1.96GiB, kv_cache_size=34.96GiB, gpu_memory_utilization=0.90.
INFO 12-13 02:10:28 worker.py:237] Memory profiling results: duration=72.36 seconds, total_gpu_memory=79.15GiB, initial_memory_usage=19.68GiB, peak_torch_memory=34.50GiB, memory_usage_post_profile=35.32GiB, non_torch_memory=1.96GiB, kv_cache_size=34.78GiB, gpu_memory_utilization=0.90.
INFO 12-13 02:10:29 distributed_gpu_executor.py:58] # GPU blocks: 189945, # CPU blocks: 21845
INFO 12-13 02:10:29 distributed_gpu_executor.py:62] Maximum concurrency for 1024 tokens per request: 2967.89x
INFO 12-13 02:10:32 model_runner.py:1409] Capturing cudagraphs for decoding. This may lead to unexpected consequences if the model is not static. To run the model in eager mode, set 'enforce_eager=True' or use '--enforce-eager' in the CLI.
INFO 12-13 02:10:32 model_runner.py:1413] If out-of-memory error occurs during cudagraph capture, consider decreasing `gpu_memory_utilization` or switching to eager mode. You can also reduce the `max_num_seqs` as needed to decrease memory usage.
(VllmWorkerProcess pid=1587462) INFO 12-13 02:10:32 model_runner.py:1409] Capturing cudagraphs for decoding. This may lead to unexpected consequences if the model is not static. To run the model in eager mode, set 'enforce_eager=True' or use '--enforce-eager' in the CLI.
(VllmWorkerProcess pid=1587462) INFO 12-13 02:10:32 model_runner.py:1413] If out-of-memory error occurs during cudagraph capture, consider decreasing `gpu_memory_utilization` or switching to eager mode. You can also reduce the `max_num_seqs` as needed to decrease memory usage.
INFO 12-13 02:10:57 custom_all_reduce.py:224] Registering 3232 cuda graph addresses
(VllmWorkerProcess pid=1587462) INFO 12-13 02:10:57 custom_all_reduce.py:224] Registering 3232 cuda graph addresses
(VllmWorkerProcess pid=1587462) INFO 12-13 02:10:57 model_runner.py:1523] Graph capturing finished in 25 secs, took 1.56 GiB
INFO 12-13 02:10:57 model_runner.py:1523] Graph capturing finished in 25 secs, took 1.56 GiB
Processed prompts: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████| 8/8 [00:06<00:00,  1.29it/s, est. speed input: 25.33 toks/s, output: 82.59 toks/s]
INFO 12-13 02:11:06 multiproc_worker_utils.py:133] Terminating local vLLM worker processes
(VllmWorkerProcess pid=1587462) INFO 12-13 02:11:06 multiproc_worker_utils.py:240] Worker exiting
INFO 12-13 02:11:20 config.py:403] This model supports multiple tasks: {'generate', 'embedding'}. Defaulting to 'generate'.
INFO 12-13 02:11:20 llm_engine.py:249] Initializing an LLM engine (v0.1.dev3512+g388ee3d) with config: model='ibm-fms/Bamba-9.8b-1.8T-hf', speculative_config=None, tokenizer='ibm-fms/Bamba-9.8b-1.8T-hf', skip_tokenizer_init=False, tokenizer_mode=auto, revision=None, override_neuron_config=None, tokenizer_revision=None, trust_remote_code=True, dtype=torch.float32, max_seq_len=1024, download_dir=None, load_format=LoadFormat.AUTO, tensor_parallel_size=1, pipeline_parallel_size=1, disable_custom_all_reduce=False, quantization=None, enforce_eager=False, kv_cache_dtype=auto, quantization_param_path=None, device_config=cuda, decoding_config=DecodingConfig(guided_decoding_backend='xgrammar'), observability_config=ObservabilityConfig(otlp_traces_endpoint=None, collect_model_forward_time=False, collect_model_execute_time=False), seed=0, served_model_name=ibm-fms/Bamba-9.8b-1.8T-hf, num_scheduler_steps=1, chunked_prefill_enabled=False multi_step_stream_outputs=True, enable_prefix_caching=False, use_async_output_proc=True, use_cached_outputs=False, mm_processor_kwargs=None, pooler_config=None,compilation_config=CompilationConfig(level=0, backend='', custom_ops=[], splitting_ops=['vllm.unified_attention', 'vllm.unified_attention_with_output'], use_inductor=True, inductor_specialize_for_cudagraph_no_more_than=None, inductor_compile_sizes=[], inductor_compile_config={}, inductor_passes={}, use_cudagraph=False, cudagraph_num_of_warmups=0, cudagraph_capture_sizes=None, cudagraph_copy_inputs=False, pass_config=PassConfig(dump_graph_stages=[], dump_graph_dir=PosixPath('.'), enable_fusion=True, enable_reshape=True), compile_sizes=[], capture_sizes=[256, 248, 240, 232, 224, 216, 208, 200, 192, 184, 176, 168, 160, 152, 144, 136, 128, 120, 112, 104, 96, 88, 80, 72, 64, 56, 48, 40, 32, 24, 16, 8, 4, 2, 1], enabled_custom_ops=Counter(), disabled_custom_ops=Counter(), static_forward_context={})
INFO 12-13 02:11:26 model_runner.py:1089] Starting to load model ibm-fms/Bamba-9.8b-1.8T-hf...
INFO 12-13 02:11:28 weight_utils.py:243] Using model weights format ['*.safetensors']
Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]
Loading safetensors checkpoint shards:  25% Completed | 1/4 [00:02<00:08,  2.85s/it]
Loading safetensors checkpoint shards:  50% Completed | 2/4 [00:05<00:05,  2.91s/it]
Loading safetensors checkpoint shards:  75% Completed | 3/4 [00:08<00:02,  2.91s/it]
Loading safetensors checkpoint shards: 100% Completed | 4/4 [00:11<00:00,  2.92s/it]
Loading safetensors checkpoint shards: 100% Completed | 4/4 [00:11<00:00,  2.91s/it]

INFO 12-13 02:11:42 model_runner.py:1094] Loading model weights took 36.6217 GB
INFO 12-13 02:11:45 worker.py:237] Memory profiling results: duration=2.65 seconds, total_gpu_memory=79.15GiB, initial_memory_usage=39.37GiB, peak_torch_memory=67.84GiB, memory_usage_post_profile=69.07GiB, non_torch_memory=2.73GiB, kv_cache_size=0.66GiB, gpu_memory_utilization=0.90.
INFO 12-13 02:11:45 gpu_executor.py:76] # GPU blocks: 1797, # CPU blocks: 10922
INFO 12-13 02:11:45 gpu_executor.py:80] Maximum concurrency for 1024 tokens per request: 28.08x
INFO 12-13 02:11:45 model_runner.py:1409] Capturing cudagraphs for decoding. This may lead to unexpected consequences if the model is not static. To run the model in eager mode, set 'enforce_eager=True' or use '--enforce-eager' in the CLI.
INFO 12-13 02:11:45 model_runner.py:1413] If out-of-memory error occurs during cudagraph capture, consider decreasing `gpu_memory_utilization` or switching to eager mode. You can also reduce the `max_num_seqs` as needed to decrease memory usage.
INFO 12-13 02:12:08 model_runner.py:1523] Graph capturing finished in 23 secs, took 0.11 GiB
Processed prompts: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████| 8/8 [00:01<00:00,  4.21it/s, est. speed input: 82.65 toks/s, output: 269.54 toks/s]
PASSED

==================================================================================================== warnings summary =====================================================================================================
../transformers/src/transformers/utils/hub.py:128
  /data/flim/transformers/src/transformers/utils/hub.py:128: FutureWarning: Using `TRANSFORMERS_CACHE` is deprecated and will be removed in v5 of Transformers. Use `HF_HOME` instead.
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================================================================================== 1 passed, 1 warning in 216.66s (0:03:36) =========================================================================================
(mamba-vllm) 1000960000@flim-mamba-master-0:~/data/vllm$

</details>

tlrmchlsmth commented on vllm/model_executor/layers/mamba/mamba_mixer2.py:
```diff
@@ -0,0 +1,494 @@
+        intemediate_settings = (intermediate_size, 0, 1)
+        head_setings = (self.num_heads, 0, 1)
+
+        delattr(self.conv1d.bias, "weight_loader")
```
Do we need to do this? I would have thought `set_weight_attrs` was sufficient.

fabianlim replied:
yea if I do not have that this will raise because i think the `weight` already has that attribute

```
File [vllm/vllm/model_executor/utils.py:29](vllm/vllm/model_executor/utils.py#line=28), in set_weight_attrs(weight, weight_attrs)
     27     return
     28 for key, value in weight_attrs.items():
---> 29     assert not hasattr(
     30         weight, key), (f"Overwriting existing tensor attribute: {key}")
     32     # NOTE(woosuk): During weight loading, we often do something like:
     33     # narrowed_tensor = param.data.narrow(0, offset, len)
     34     # narrowed_tensor.copy_(real_weight)
   (...)
     39     # we sync the param tensor after its weight loader is called.
     40     # TODO(woosuk): Remove this hack once we have a better solution.
     41     if current_platform.is_tpu() and key == "weight_loader":

AssertionError: Overwriting existing tensor attribute: weight_loader
```

fabianlim replied:
I added a comment regarding the need for `delattr` [here](https://github.com/vllm-project/vllm/pull/10909#:~:text=%E2%80%A6-,400db27,-Merge%20state)

tlrmchlsmth commented on vllm/model_executor/layers/mamba/ops/ssd_bmm.py:
```diff
@@ -0,0 +1,267 @@
+# Adapted from https://github.com/state-spaces/mamba/blob/main/mamba_ssm/ops/triton/ssd_bmm.py
+
+# ruff: noqa: E501,SIM102
+"""We want triton==2.1.0 or 2.2.0 for this
+"""
```
I think this comment is out-of-date and should be deleted (Likely triton will be >= 3.1.0)

ditto for other files

fabianlim replied:
Done in https://github.com/vllm-project/vllm/pull/10909/commits/d788db694330d27ffc0f269dfbb1ccef3eb82f72, also removed softplus.py and reused the one already in mamba_ssm.py

tlrmchlsmth reviewed (commented):
first pass, just a few comments. At a high level it looks good. 

Will you add a test for tensor parallelism?

fabianlim committed:
reuse softplus and remove triton2 remark

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
add comment on weight loader and format

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
rename test_jamba to test_hybrid and got rid of test_bamba

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim commented:
@tlrmchlsmth i have addressed most of your comments now, not rebasing yet, waiting for you to look first. But I realized `test_jamba.py` has changed so I will need to do the rename and test again.

tlrmchlsmth reviewed (commented):
@fabianlim At a high level, the changes look good, and the PR looks good overall. I'll do a more thorough review once it's unmarked as draft.

Could you add unit tests for the added kernels in `layers/mamba/ops`?

fabianlim committed:
Merge remote-tracking branch 'upstream/main' into bamba-pr

fabianlim committed:
update bamba to ishybrid and support pp

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
lint

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
add unit test for mamba ssd

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
fix lint

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
full chunked-prefill fix (sans unit tests)

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
format and add cont batch unit tests (will need more cases)

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
fix kernel tests and add more chunked prefill cases

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
bound adjustment

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim commented:
@tlrmchlsmth I have now marked the PR as ready and have addressed the remaining items, in particular the chunked prefill is the biggest change. This requires changes into the kernels. To test this, I have added unit tests for various chunked prefill scenarios. 

Know is the holiday season and pls take your time. Happy holidays!

fabianlim committed:
bound adjustment

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
lint errors

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

tlrmchlsmth commented on vllm/model_executor/layers/mamba/mamba_mixer2.py:
```diff
@@ -0,0 +1,499 @@
+        # - HOWEVER IF, world_size DOES NOT divide groups, then we need
+        #   to allocate extra space in the shard, such that groups
+        #   may be replicated to follow the head shard.
+        self.tp_size = get_tensor_model_parallel_world_size()
+        tp_rank = get_tensor_model_parallel_rank()
```
Here as well it would be good to add asserts to check that the tensor parallel world size is valid

tlrmchlsmth commented on vllm/model_executor/layers/mamba/mamba_mixer2.py:
```diff
@@ -0,0 +1,499 @@
+            self.num_heads //
+            n_groups,  # ratio for mapping back to original group
+        )
+        intemediate_settings = (intermediate_size, 0, 1)
```
nit: 
```suggestion
        intermediate_settings = (intermediate_size, 0, 1)
```

fabianlim replied:
renaming is best done with refactoring tool, so I did not accept the commit suggestion

tlrmchlsmth reviewed (commented):
@fabianlim do you know when a version of transformers that supports bamba will be released? This PR will likely be blocked until it's out

tlrmchlsmth commented on vllm/model_executor/models/bamba.py:
```diff
@@ -0,0 +1,610 @@
+        self.rope_theta = rope_theta
+        self.max_position_embeddings = max_position_embeddings
+
+        self.rotary_emb = RotaryEmbedding(
+            head_size=self.head_dim,
+            rotary_dim=config.attn_rotary_emb,
+            max_position_embeddings=max_position_embeddings,
+            base=rope_theta,
+            is_neox_style=True,
+            dtype=torch.get_default_dtype(),  # see impl of get_rope
+        )
```
Why does `bamba` construct the `RotaryEmbedding` directly?  I think it can use `get_rope` instead

tlrmchlsmth commented on vllm/model_executor/models/bamba.py:
```diff
@@ -0,0 +1,610 @@
+        qkv, _ = self.qkv_proj(hidden_states)
+        q, k, v = qkv.split([self.q_size, self.kv_size, self.kv_size], dim=-1)
+
+        # because the bamba model may potentially handle long sequences,
+        # we should adjust the sin_cos cache if necessary to avoid out of bounds
+        # - first get the max_position
+        max_position = max(
+            getattr(attn_metadata, 'max_prefill_seq_len', 0),
+            getattr(attn_metadata, 'max_decode_seq_len', 0),
+        )
+        if max_position == 0:
+            # if we cannot get the max length from the metadata, then
+            # get it from the positions
+            max_position = positions.max().item()
+
+        # when VLLM_ALLOW_LONG_MAX_MODEL_LEN=1 could potentially cause inputs
+        # longer than max_position_embeddings. We extend the rope cache
+        # to prevent CUDA errors. Be aware that the outputs could be of
+        # lower quality for long sequence lengths.
+        rotary = self.rotary_emb
+        if rotary.max_position_embeddings <= max_position:
+            # we set it to the next power of two that covers it
+            while rotary.max_position_embeddings <= max_position:
+                rotary.max_position_embeddings *= 2
+            rotary.cos_sin_cache = rotary._compute_cos_sin_cache()
```
Instead of this, have you considered using `rope_scaling` instead? If you use `get_rope` instead of constructing the RotaryEmbedding directly, I think it should work for `bamba`

See this unit test for an example of how it works:
https://github.com/sasha0552/vllm/blob/d427e5cfda8d2536b81e6021128e71b2dbc281aa/tests/test_config.py#L177

See also:
https://www.reddit.com/r/LocalLLaMA/comments/14mrgpr/dynamically_scaled_rope_further_increases/

tlrmchlsmth commented on vllm/model_executor/layers/mamba/mamba_mixer2.py:
```diff
@@ -0,0 +1,499 @@
+        seq_len, _ = hidden_states.shape
+        groups_time_state_size = self.n_groups * self.ssm_state_size
+
+        # - doing it differently from mixer v1; little confused with its logic
+        # - we need to do is to detect if there is any prefill; if there are
+        #   no prefils, then each example will be coming in one sample at a time
+        # - on the other hand v1 checks for "query_start_loc"
+        #   and "context_lens_tensor" however we have noticed that, even
+        #   when the samples are coming in
+        #   one at a time, they are still not NONE, e.g.,
+        #   * "query_start_loc" = [0, 1, ..]
+        #   * "context_lens_tensor" = [8, ...]
```
This does appear to be an issue in `MambaMixer`'s `forward` method. You can go ahead and delete this comment, and I'll follow up with a PR to fix what's happening there

fabianlim commented on tests/kernels/test_mamba_ssm_ssd.py:
```diff
@@ -0,0 +1,283 @@
+
+# Added by the IBM Team, 2024
+
+# Adapted from https://github.com/state-spaces/mamba/tree/main/mamba_ssm/ops/triton
```
ah ok.. just to confirm, this is a permalink `Adapted from https://github.com/state-spaces/mamba/blob/v2.2.4/mamba_ssm/modules/ssd_minimal.py`. I have seen those with hashes in the URL before, you are not referring to those?

fabianlim commented:
hey @tlrmchlsmth thanks for your comments will get to them, the model has been merged to HF main, but it is still pending an official release to my knowledge. Hopefully that should happen soon since the last patch release was 2 weeks ago.

fabianlim committed:
Add permalink correction from @tlrmchlsmth

Co-authored-by: Tyler Michael Smith <tyler@neuralmagic.com>
Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
improved comment for segsum, add more sizes for test_mamba_chunk_scan_single_example

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
rename and comment functions, add more sizes for test_mamba_chunk_scan_cont_batch

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
addressed comments on mamba_mixer2.py

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
replace with get_rope

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
rope scaling

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
fixes

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
zero out ssm states

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
fix tests (sans updating dev checkpoint)

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
not replacing dev model for now

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim commented:
@tlrmchlsmth [transformers v4.48 ](https://github.com/huggingface/transformers/releases/tag/v4.48.0) has just been released and along with it Bamba. So maybe we can consider merging this PR first as the dev model might take some more time.

tlrmchlsmth reviewed (commented):
Could you update transformers to 4.48.0 in `requirements-common.txt`? (And note that it's required for bamba)

tlrmchlsmth reviewed (commented):
I've found that `test_hybrid.py` is red on an H100

```
pytest tests/models/decoder_only/language/test_hybrid.py::test_mamba_prefill_chunking -vs -k Bamba
```
Results in
```
E           AssertionError: Test1:
E           chunked:    'Briefly describe the major milestones in the development of artificial intelligence from 1950 to 2020.\nArtificial intelligence (AI) has been a topic'
E           non_chunked:        'Briefly describe the major milestones in the development of artificial intelligence from 1950 to 2020.\nArtificial intelligence (AI) is a field of'
```

PTAL - I think we could relax the test to fix this issue

fabianlim committed:
update requirements

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
remove extraneous comment

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
update test

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim commented:
@tlrmchlsmth I have updated the test by changing the `max_tokens` down to 7 and adding a `dype` change, it should be more relaxed now.

tlrmchlsmth reviewed (approved):
Looks good once the linter is green. Thanks for the great work!

fabianlim committed:
fix lint

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
fix lint

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim commented:
@tlrmchlsmth we have already updated `requirements-common.txt` to upgrade to `transformers==4.48.0` so that we have the new model, but i noticed that `requirements-test.txt` also has a `transformers` version pinned. That is what is causing the [failure](https://buildkite.com/vllm/ci/builds/11978#0194674e-7e48-4a7c-b9d3-f995018729f3/105-7654).

fabianlim committed:
fix requirements-test

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
update test registry, fixes

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

tlrmchlsmth committed:
Mamba2 changes from #10909

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>
Co-authored-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

tlrmchlsmth commented on vllm/model_executor/layers/mamba/mamba_mixer2.py:
```diff
@@ -0,0 +1,493 @@
+            "Tensor parallel world size must divide num heads."
+
+        self.ssm_state_size = ssm_state_size
+        self.use_rms_norm = use_rms_norm
```
Just noticed that this doesn't seem to be used at all. Should it be removed, or is it to be used somewhere?

fabianlim replied:
oh yea.. let me check and I can remove

tlrmchlsmth committed:
Get Mamba2 working!

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
Add integration test -- something is wrong!!

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
format

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
fixes

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
Fix for conv state shape and update placeholder_attn

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>
Co-authored-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

tlrmchlsmth committed:
back out placeholder_attn changes

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

fabianlim committed:
make seq_idx to chunk indices more efficient

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

tlrmchlsmth committed:
WIP debugging, restore local mamba and placeholder_attn changes

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
Integration tests are now green

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
remove bamba-specific files

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

yury-tokpanov commented on vllm/model_executor/models/bamba.py:
```diff
@@ -0,0 +1,599 @@
+        lora_config = vllm_config.lora_config
+
+        self.config = config
+        self.padding_idx = config.pad_token_id
```
is this one used anywhere?

fabianlim replied:
oh sorry good catch I will remove it

tlrmchlsmth committed:
Merge branch 'main' into tms/mamba2

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
Handle grouping in Mixer2RMSNormGated

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
debug cruft

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

tlrmchlsmth committed:
Remove codestral integration test

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

fabianlim commented:
Note to self, some of the testing API has changed due to this PR https://github.com/vllm-project/vllm/pull/10353

fabianlim committed:
Merge branch 'tms/mamba2' into bamba-pr

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
update mamba_cache

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
remove changes to requirements

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

tlrmchlsmth commented on vllm/model_executor/models/mamba2.py:
Let's land mamba 2 in #9292

tlrmchlsmth commented on requirements-common.txt:
Could you merge in latest main? We've already landed this change

fabianlim replied:
ya actually yest i reverted this file and took it from latest main, but somehow the diff shows up in github. The version on the left shown by github is actually old

fabianlim replied:
ok if i merge in latest main it seems fine..

tlrmchlsmth commented on requirements-test.txt:
I don't think this file should be changed now that https://github.com/vllm-project/vllm/pull/12599 has been merged. Could you revert this file?

Also wondering why this has more changes than in #12599 - did you run into any additional issues that required these additional chagnes?

fabianlim replied:
same as above, the version on the left is old.

tlrmchlsmth commented on tests/models/decoder_only/language/test_mamba.py:
Let's land these changes as part of #9292

tlrmchlsmth commented on vllm/model_executor/layers/mamba/mamba_mixer2.py:
```diff
@@ -0,0 +1,532 @@
+            if redundant_tp:
+                # To handle the general case, redundantly apply the variance
+                x = tensor_model_parallel_all_gather(x, -1)
+
+            *prefix_dims, hidden_dim = x.shape
+            group_count = hidden_dim // self.group_size
+            x_grouped = x.view(*prefix_dims, group_count, self.group_size)
+            variance = x_grouped.pow(2).mean(-1, keepdim=True)
+            x_grouped = x_grouped * torch.rsqrt(variance +
+                                                self.variance_epsilon)
+            x = x_grouped.view(*prefix_dims, hidden_dim)
+
+            if redundant_tp:
+                start = self.per_rank_hidden_size * self.tp_rank
+                end = start + self.per_rank_hidden_size
+                x = x[..., start:end]
+
+        return self.weight * x.to(input_dtype)
+
+    def forward_cuda(
+        self,
+        x: torch.Tensor,
+        gate: torch.Tensor,
+    ) -> Union[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]:
+
+        if self.tp_size > 1 or self.n_groups != 1:
+            return self.forward_native(x, gate)
+
+        from vllm import _custom_ops as ops
+
+        # cast x and gate to float32 before silu
+        out = torch.empty_like(x)
+        y = x * nn.functional.silu(gate.to(torch.float32))
+        ops.rms_norm(
+            out,
+            y.to(x.dtype),
+            self.weight.data,
+            self.variance_epsilon,
+        )
+        return out
```
We should get a unit test in place for this, especially the various tensor parallel cases. @fabianlim do you have bandwidth to do that? Otherwise I can do it in either in #9292 or a separate PR. I do feel pretty good about correctness here, having manually tested various cases thoroughly enough.

fabianlim replied:
just unit testing only the `Mixer2RMSNormGated`? if so how would you setup the test? `conftest` only has runners for the whole model.

tlrmchlsmth replied:
Thanks for adding the test. Here is good!

tlrmchlsmth commented on vllm/model_executor/models/registry.py:
```diff
@@ -69,6 +70,7 @@
     "LLaMAForCausalLM": ("llama", "LlamaForCausalLM"),
     "MambaForCausalLM": ("mamba", "MambaForCausalLM"),
     "FalconMambaForCausalLM": ("mamba", "MambaForCausalLM"),
+    "Mamba2ForCausalLM": ("mamba2", "Mamba2ForCausalLM"),
```
Let's land this in #9292

tlrmchlsmth reviewed (commented):
Looking pretty good, let's get this landed now that 4.48.2 is out!

tlrmchlsmth commented:
To debug the pre-commit issue locally you may need to run:
```
pre-commit run mypy-3.9 --hook-stage manual -a
```

fabianlim committed:
revert changes

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
Merge remote-tracking branch 'upstream/main' into bamba-pr

fabianlim committed:
more reverts

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
fix lint

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
fix lint

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim commented:
@tlrmchlsmth thank you so much for your comments. I have fixed the `pre-commit` and reverted changes in a bunch of files; i kept the changes that are still needed to test the PR. Also merging in upstream/main helped github to display the changes correctly. Regarding the new unit test, I left a question for you

fabianlim committed:
remove unnecessary stuff

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
add mixer2 gated norm TP test

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
Merge remote-tracking branch 'upstream/main' into bamba-pr

fabianlim committed:
add header

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
fix lint

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

tlrmchlsmth committed:
Merge branch 'main' into bamba-pr

tlrmchlsmth commented:
I've been working on getting this PR green. Some of the issues are due to huggingface timeouts, but the kernels-test-3 failure is real. The log is too large to link to the problematic lines, but there is an illegal memory access during test_mamba_chunk_scan_cont_batch.

tlrmchlsmth commented on tests/models/decoder_only/language/test_hybrid.py:
```diff
@@ -8,7 +8,8 @@
 
-MODELS = ["ai21labs/Jamba-tiny-dev"]
+# This test is for the hybrid models
+MODELS = ["ai21labs/Jamba-tiny-dev", "ibm-fms/Bamba-9B"]
```
Reminder to change before landing
```suggestion
MODELS = ["ai21labs/Jamba-tiny-dev", "ibm-ai-platform/Bamba-9B"]
```

tlrmchlsmth commented on tests/models/registry.py:
```diff
@@ -102,6 +102,7 @@ def check_available_online(
                                          trust_remote_code=True),
     "BaichuanForCausalLM": _HfExamplesInfo("baichuan-inc/Baichuan2-7B-chat",
                                          trust_remote_code=True),
+    "BambaForCausalLM": _HfExamplesInfo("ibm-fms/Bamba-9B"),
```
ditto

fabianlim committed:
checkpoint renames

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim commented:
@tlrmchlsmth oh man i cant reproduce it in my machine

```
python -m pytest -x tests/kernels/test_mamba_ssm_ssd.py::test_mamba_chunk_scan_cont_batch
============================================== test session starts ==============================================
platform linux -- Python 3.9.19, pytest-8.3.4, pluggy-1.5.0
rootdir: /gpfs/users/flim/data/vllm
configfile: pyproject.toml
plugins: anyio-4.8.0
collected 192 items

tests/kernels/test_mamba_ssm_ssd.py ..................................................................... [ 35%]
......................................................................................................... [ 90%]
..................                                                                                        [100%]

======================================= 192 passed in 2713.89s (0:45:13) ========================================
```

fabianlim committed:
(debug) test_mamba_ssm_ssd.py

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
[debug] make all run same shard_id

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

fabianlim committed:
[debug] disable test case

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

tlrmchlsmth commented on vllm/model_executor/layers/mamba/ops/ssd_chunk_scan.py:
```diff
@@ -0,0 +1,615 @@
+        seq_idx_ptr += pid_b * stride_seq_idx_batch + c_idx * chunk_size * stride_seq_idx_seqlen
+
+        # - seq_idx_prev points to be previous (possibly logical) chunk.
+        seq_idx_prev = tl.load(seq_idx_ptr - stride_seq_idx_seqlen,
+                               mask=pid_c >= 1,
+                               other=0)
```
should it be this?
```suggestion
        seq_idx_prev = tl.load(seq_idx_ptr - stride_seq_idx_seqlen,
                               mask=c_idx >= 1,
                               other=0)
```

fabianlim replied:
because we need the `seq_idx` from the previous logical chunk here. say the second chunk has been split into two, and we are at the splitted (micro) chunk, then pid_c = 2, c_idx = 1 for this chunk, and we want the previous (mirco) chunk, which is pid_c == 1

| chunk 0                             | chunk 1 ---      | split                      |
| pic_c = 0                           | pic_c = 1          | split (pic_c == 2. |

But yes I think the there is some edge case here, if pid_c == 1, and if c_idx == 0, then `c_idx` is at the "edge" and we cannot decrease the pointer like this

fabianlim replied:
yes i think i might have overlooked this.  `seq_idx_prev` is mostly used only in the case `HAS_INITSTATES==False` and in the case `HAS_INITSTATES==True` but `c_off==0`, which means  `seq_idx_ptr` should only be positinoed with reference to chunk boundaries, hence we should be using `c_idx` to check for the mask

fabianlim committed:
revert debugs and add @tlrmchlsmth fix!

Signed-off-by: Yu Chin Fabian Lim <flim@sg.ibm.com>

tlrmchlsmth committed:
Merge branch 'main' into bamba-pr

tlrmchlsmth committed:
update mamba and jamba for MambaCache changes

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>
